openapi/spec-generator/specbuilder.go

```diff
@@ -0,0 +1,490 @@
+// adds all referenced schemas to the schemas map, but DOES NOT add the request
+// or response schema. The callers are responsible for adding the root to the
+// correct map.
+func (b *SpecBuilder) getOrCreateSchemaRef(model reflect.Type) (*openapi3.SchemaRef, error) {
+	var ok bool
+	var err error
+	var ref *openapi3.SchemaRef
+	// if it doesn't exist generate
+	if ref, ok = b.spec.Model.Components.Schemas[model.Name()]; !ok {
+		ref, err = b.Generator.GenerateSchemaRef(model)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Add all generated schema refs if not already added
+	for schemaRef, _ := range b.Generator.SchemaRefs {
+		if b.isBasic(schemaRef.Ref) {
+			continue
+		}
+		if len(schemaRef.Ref) > 0 && !strings.Contains(schemaRef.Ref, "#/components/schemas") {
+			if _, ok = b.spec.Model.Components.Schemas[schemaRef.Ref]; !ok {
+				b.spec.Model.Components.Schemas[schemaRef.Ref] = openapi3.NewSchemaRef("", schemaRef.Value)
+			}
+		}
+	}
+
+	return ref, nil
+}
+
+func (b *SpecBuilder) isBasic(typ string) bool {
+	return typ == "" || typ == "integer" || typ == "number" || typ == "string" || typ == "boolean"
+}
+
+func (b *SpecBuilder) resolveRefPaths() {
+	for _, schemaRef := range b.spec.Model.Components.Schemas {
+		// Next make sure the refs point to other schemas, if not already done.
+		for _, property := range schemaRef.Value.Properties {
+			if strings.Contains(property.Ref, "#/components/schemas/") {
+				continue
+			} else if b.isBasic(property.Value.Type) {
+				property.Ref = ""
+			} else if property.Value.Type == "array" {
+				if b.isBasic(property.Value.Items.Value.Type) {
+					property.Value.Items.Ref = ""
+				} else {
+					if !strings.Contains(property.Value.Items.Ref, "#/components/schemas") {
+						property.Value.Items.Ref = fmt.Sprintf("#/components/schemas/%s", property.Value.Items.Ref)
+					}
+				}
+			} else if property.Value.AdditionalProperties != nil {
+				property.Ref = ""
+				if !b.isBasic(property.Value.AdditionalProperties.Value.Type) {
+					if len(property.Value.AdditionalProperties.Ref) > 1 {
+						if !strings.Contains(property.Value.AdditionalProperties.Ref, "#/components/schemas") {
+							property.Value.AdditionalProperties.Ref = fmt.Sprintf("#/components/schemas/%s", property.Value.AdditionalProperties.Ref)
+						}
+					}
+				}
+			} else {
+				if !strings.Contains(property.Ref, "#/components/schemas") {
+					property.Ref = fmt.Sprintf("#/components/schemas/%s", property.Ref)
+				}
+			}
+		}
+	}
+}
+
+func (b *SpecBuilder) adaptSecurityRequirements(parameters []*Parameter) *openapi3.SecurityRequirements {
+	var tokenParam *Parameter
+	for _, param := range parameters {
+		if param.Name == "X-Nomad-Token" {
+			tokenParam = param
+			break
+		}
+	}
+
+	if tokenParam == nil {
+		return nil
+	}
+
+	var ok bool
+	var securitySchemeRef *openapi3.SecuritySchemeRef
+	if securitySchemeRef, ok = b.spec.Model.Components.SecuritySchemes[tokenParam.Name]; !ok {
```

Comment:
Do you need to declare this? Would the following work?
```suggestion
	if securitySchemeRef, ok := b.spec.Model.Components.SecuritySchemes[tokenParam.Name]; !ok {
```